### DISCUSSION TITLE
Cast to two interfaces

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/csharplang#4344] [Q&A] [answered]
I have a generic type that I cannot control that implements multiple interfaces (lets call it `Source`). And I have a wrapper class `Target` (that I control) that manipulates members of two non-generic interfaces (that `Source` implements). I need to create a variable from instance of `Source` class that I can pass to constructor of `Target` class.

So something like this. It would be great if `t is D d` returned `true`, but it doesn't..

```cs
public interface A { }
public interface B { }
public interface C<T> { }

public interface D : A, B { }

public class Source<T> : A, B, C<T> { }
public class Target<T> where T : A, B { /*Constructor and usage of A and B members here*/  }

static void Main()
{
    Source<int> s = new();
    Converter(s);
}

static void Converter(object obj)
{
    if (obj is D d) // false
    {
        new Target(d);
    }
    else if (obj is A && obj is B) // true
    {
        new Target(???);
    }
}
```

PS: Method where this logic is located accepts `object` (that may be `Source` instance) and I cannot change that, so I don't know what type argument instance of `Source` has.

**Comment by Joe4evr** [comment]
At first it sounds like you want Shapes/Roles (#164/#1711), but in playing around a bit with your sample [you really don't need any of that to get it working](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFCEDMABJgHbwLYQDGcdAkgIJ0BvOgF8qtBs0RtOPAEKCRY+kxbSu3AMIAeACoA+BaMpK6JOgGUA9gFcEnXQZA9eSOa80PDJszogIA5nAwngDuABaIXDp0TnzuslQCVHQppvS+AUEAFNGY8GAAlF6URmKkAGymxHQaicmphBWmACx0ALIQTFkF9SlJlKmDdGh+dFB0ALx0jHAhFjZ2cFoqet0A3L1DIwh0MJPTs3QZgcFWtvYrq1AFGwOpRsJAA=).

**Reply by maxkoshevoi** [reply]
Thanks for the answer! I'll keep a look on those features. 

One more thing. Method where this logic is located accepts `object` (that may be `Source` instance) and I cannot change that, so I don't know what type argument instance of `Source` has.

The only way I found to achieve that is to cast `t` to `A` and to `B` separately and then pass two resulting variables into `new Target(a, b)`

**Reply by BreyerW** [reply]
What you are looking for is called union&intersection types (eg `IA&IB` or `IA|IB`) but c# unfortunetly doesnt support them yet

**Comment by bbarry** [comment, marked as answer]
This compiles:

```csharp
public interface A { }
public interface B { }
public interface C<T> { }

public interface D : A, B { }

public class Source<T> : A, B, C<T> { }
public class Target<T> where T : A, B { 
    readonly T _value;
    public Target(T value) {
        _value = value;
    }
    public void DoWork() {}
}

public static class TargetCreator {
    public static Target<T> Create<T>(T value) where T : A, B => new Target<T>(value);
}

public class M {
    static void Main()
    {
        Source<int> t = new();
        if (t is D d) // false
        {
            TargetCreator.Create(d).DoWork();
        }
        else if (t is A && t is B) // true
        {
            TargetCreator.Create(t).DoWork();
        }
    }
}
```

**Reply by maxkoshevoi** [reply]
Sorry, I didn't clarify one thing (updated description). Method where this logic is located accepts `object` (that may be `Source` instance) and I cannot change that, so I don't know what type argument instance of `Source` has.

**Reply by bbarry** [reply]
In that case you have to trick the compiler a little:

```csharp

public class M {
    static void Accept(object t)
    {
        if (t is D d) // false
        {
            TargetCreator.Create(d).DoWork();
        }
        else if (t is A && t is B) // true
        {
            TargetCreator.Create((dynamic)t).DoWork();
        }
    }
}
```

**Reply by BreyerW** [reply]
Using `dynamic` is not good idea since it pulls in dynamic machinery and performance is lower too. Avoid it if you can.

**Comment by gulshan** [comment]
See #399 Intersection types for interfaces.
